src/transformers/tokenization_utils_fast.py

```diff
@@ -863,3 +864,99 @@ def train_new_from_iterator(
             kwargs["additional_special_tokens"] = additional_special_tokens
 
         return self.__class__(tokenizer_object=tokenizer, **kwargs)
+
+    def _update_normalizer(self):
+        """Updates the underlying normalizer with the current `add_prefix_space` and `legacy` settings."""
+        sequence = json.loads(normalizers.Sequence([]).__getstate__())
+        final_sequence = normalizers.Sequence([])
+        if self._tokenizer.normalizer is not None and type(self._tokenizer.normalizer) in (
+            normalizers.Prepend,
+            normalizers.Precompiled,
+        ):
+            # If normalizer is not a Sequence, add it to a sequence
+            sequence["normalizers"].append(json.loads(self._tokenizer.normalizer.__getstate__().decode("utf-8")))
+        elif type(self._tokenizer.normalizer) is normalizers.Sequence:
+            curr_state = json.loads(self._tokenizer.normalizer.__getstate__().decode("utf-8"))
+            # Remove Prepend normalizer if it exists as we are updating it
+            sequence["normalizers"] = [pt for pt in curr_state["normalizers"] if pt["type"] not in ["Prepend"]]
+        else:
+            return
+        if getattr(self, "legacy", True):
+            if getattr(self, "add_prefix_space", True):
+                new_prepend = json.loads(normalizers.Prepend(prepend="▁").__getstate__().decode("utf-8"))
+                sequence["normalizers"].append(new_prepend)
+            if not any(n["type"] == "Replace" for n in sequence["normalizers"]):
+                new_replace = json.loads(normalizers.Replace(pattern=" ", content="▁").__getstate__().decode("utf-8"))
+                sequence["normalizers"].append(new_replace)
+            final_sequence.__setstate__(json.dumps(sequence).encode("utf-8"))
+            self._tokenizer.normalizer = final_sequence
+            return False
+        if not getattr(self, "legacy", True):
+            self._tokenizer.normalizer = final_sequence
+            return
+
+    def _update_pre_tokenizer(self):
+        """Updates the underlying pre-tokenizer with the current `add_prefix_space` setting."""
+
+        # 'add_prefix_space' not passed, try to read from normalizer, otherwise do not set.
```

Comment:
Here we should also first check if we have `legacy` set to `False`. If it is `True` we return without doing anything. If `legacy` is set to `False`, then we can check if `add_prefix_space` was passed or not. 

If it is not passed, then we :
- remove prepend if prepend was in the sequence. I think shipping native pythonic way of doing this is gonna be a priority for me 
- add the MetaspaceTokenizer if we had the PrependNormalizer
- set prepend to "first' if we had metaspace
- set add_prefix_space of the bytelevel tokenizer to true / false if there was one 
- do the same of the decoder (if we update the pre_tokenizer we need to update the decoder as well)
-

Reply:
@ArthurZucker  Thank you for the detailed review!
 
 Q.
 - _set prepend to "first' if we had metaspace_ - based on `convert_slow_tokenizer`, prepend is `always` only if `legacy=True`. So here should it be updated to always set to `first` ? (if we are not touching legacy)
 
 - _do the same of the decoder (if we update the pre_tokenizer we need to update the decoder as well)_ - what if there was no pretokenizer and a Metaspace one was added? do we need to add a ByteLevel decoder ?(if yes - for which cases (`prepend _scheme=first` and/or `never`)

Reply:
- if legacy:
    - do nothing 
- else:
    - if there is a MetaSpace:
        - set prepend to "first" if `add_prefix_space`
        - set prepend to "never" if not `add_prefix_space`
    - if there is a prepend Normalizer:
        - we could check the type, mostly Llama is the one using it 
        - replace with a MetaSpace, with either first or never

Reply:
Thank you! Applied in next review!

notes on the decoder:
- decoder's `add_prefix_space` is updated if decoder is bytelevel 
- otherwise leave it
- if there is no decoder, I do not add one